modules/messages/src/lib.rs

```diff
@@ -305,6 +308,36 @@ pub mod pallet {
 				payload: encoded_payload,
 				fee: delivery_and_dispatch_fee,
 			});
+			// Guaranteed to be called outside only when the message is accepted.
+			// We assume that the maximum weight call back used is `single_message_callback_overhead`, so do not perform
+			// complex db operation in callback. If you want to, put these magic logic in outside pallet and control
+			// the weight there.
+			let single_message_callback_overhead = T::WeightInfo::single_message_callback_overhead(T::DbWeight::get());
+			let actual_callback_weight = T::OnMessageAccepted::on_messages_accepted(&lane_id, &nonce);
+			match single_message_callback_overhead.checked_sub(actual_callback_weight) {
+				Some(difference) if difference == 0 => (),
+				Some(difference) => {
+					log::trace!(
+						target: "runtime::bridge-messages",
+						"Messages accepted callback has returned unspent weight to refund the submitter: \
+						{} - {} = {}",
+						single_message_callback_overhead,
+						actual_callback_weight,
+						difference,
+					);
+					actual_weight -= difference;
+				}
+				None => {
+					debug_assert!(false, "The accepted callback is wrong");
+					log::error!(
+						target: "runtime::bridge-messages",
+						"Messages accepted callback has returned more weight that it may spent: \
+						{} vs {}",
+						single_message_callback_overhead,
+						actual_callback_weight,
+					);
```

Comment:
I just keep this case the same as what `OnDeliveryConfirmed` does. But, I think you are right, it would be better to return with an error here. @svyatonik Is there any side effect here if we exit with an error directly?

Reply:
1) you can't simply fail here, because storage has been modified already. You may wrap the whole code in transaction, though;
2) the only way to reach this line is to execute it in invalid runtime. What's better way to handle invalid runtime - to fail transaction (means that the submitter loses his fee he has paid for the tx + the lane is blocked) or to simply log an error? The similar case is when extrinsic weight formula returns something that literally says `I'll be executing for 1 second`, but actually it is executing for 2 seconds. We don't fail the tx in this case && the only difference is that here we log the error. So imo let's leave it as is

Reply:
I recheck the logic in the send_message call, maybe we can not directly return an error in this case, since the new message nonce has been generated and OutboundLaneData has been changed when the callback is invoked. If we directly return an error in this case, then we change lane important storage and return error finally, that broken substrate dispatch call rule。